repos/system_upgrade/common/actors/checkfirstpartitionoffset/libraries/check_first_partition_offset.py

```diff
@@ -0,0 +1,96 @@
+    except CalledProcessError as err:
+        # Unlikely - if the disk has no partition table, `fdisk` terminates with 0 (no err). Fdisk exits with an err
+        # when the device does not exists, or if it is too small to contain a partition table.
+
+        err_msg = 'Failed to run `fdisk` to obtain the partition table of the device {0}. Full error: \'{1}\''
+        api.current_logger().error(err_msg.format(device, str(err)))
+        raise
+
+    table_iter = iter(partition_table)
+
+    for line in table_iter:
+        if not line.startswith('Units'):
+            # We are still reading general device information and not the table itself
+            continue
+
+        unit = line.split('=')[2].strip()  # Contains '512 bytes'
+        unit = int(unit.split(' ')[0].strip())
+        break  # First line of the partition table header
+
+    for line in table_iter:
+        line = line.strip()
+        if not line.startswith('Device'):
+            continue
+
+        part_all_attrs = split_on_space_segments(line)
+        break
+
+    partition_offsets = []
+    for partition_line in table_iter:
+        # Fields:               Device     Boot   Start      End  Sectors Size Id Type
+        # The line looks like: `/dev/vda1  *       2048  2099199  2097152   1G 83 Linux`
+        part_info = split_on_space_segments(partition_line)
+
+        # If the partition is not bootable, the Boot column might be empty
+        part_start = int(part_info[2]) if len(part_info) == len(part_all_attrs) else int(part_info[1])
+        partition_offsets.append(part_start)
+
+    if not partition_offsets:
+        raise ValueError('The disk {0} contains no partitions'.format(device))
+
+    first_partition_offset = min(partition_offsets)
+    offset_size = first_partition_offset * unit
+
+    return offset_size >= SAFE_OFFSET_BYTES
+
+
+def check_first_partition_offset():
+    devices_to_check = grub.get_grub_devices()
+
+    problematic_devices = []
+    for device in devices_to_check:
+        try:
+            is_ok = is_first_partition_offset_ok(device)
+            if not is_ok:
+                problematic_devices.append(device)
+        except CalledProcessError:
+            # The path does not point to an actual device, or the disk does not exists - grub.get_grub_devices lied?
+            msg = 'Failed to check the offset of the first partition of the device {0}'
+            raise StopActorExecutionError(msg.format(device))
+
+    if problematic_devices:
+        summary = ('GRUB devices with too little space reserved before the first partition were found. '
+                   'Such devices can present a problem when updating the boot loader. Problematic devices: \n'
+                   '{0}')
+        problematic_devices_fmt = ['- {0}'.format(dev) for dev in problematic_devices]
+
+        hint = 'Reformat the devices so that there is at least {0} kiB space before the first partition.'
```

Comment:
In case of summary, it would be nice to be specific, that failed update of the bootloader could lead to unbootable system, so they are not asking for the explanation what could happen. Also, we should specify clearly, that for in-place upgades it's required to start the first partition on sector 2048 or higher, to keep 1MB space before the first partition.

We will need to update this to be more specific. Especially the hint, as
* such an operation is not officially supported and customers are doing it on their own risk
* in some cases it does not have to be possible to do such an operation at all (e.g. if they used XFS for all partitions,  small amount of space on /boot already and shift of several partitions is problematic..., etc)
* also for number of our customers it's out of their skills to do something like that

So in this case, we should suggest also new clean installation of the target RHEL system with new partition layout.

Reply:
todo(@pirat89 ): check whether KCS exists, in such a case we should add the link here too